### ISSUE TITLE
Feature Request: HTTP proxy

### ISSUE THREAD

**Issue by targence** [shadowsocks/go-shadowsocks2#27]
Hi! Please add ability run SOCK and HTTP proxy at the same time.

**Comment by riobard** [comment]
Any idea why? This project is under the umbrella shadow<em>socks</em>, so HTTP proxy is kinda beyond the scope.

**Comment by daciwei** [comment]
Hi

HTTP proxy **is needed on client side**, because not all apps support SOCKS.
If we are talking about apps in terminal, there almost no apps which supports SOCKS.
I use command line apps for development a lot. 

I use ShadowsocksX-NG as client.
https://github.com/shadowsocks/ShadowsocksX-NG
There is option HTTP proxy , it's very handy.

![screen shot 2017-04-11 at 11 09 05 am](https://cloud.githubusercontent.com/assets/13871989/24899264/4e1b065e-1ea8-11e7-928c-68d14a939c85.png)

So I think it's not beyond the scope at all. HTTP proxy on client side is needed if you want to use not only browser.

**Comment by riobard** [comment]
> If we are talking about apps in terminal, there almost no apps which supports SOCKS.

This is incorrect. Lots of terminal utilities support SOCKS, e.g. curl.

> I use ShadowsocksX-NG as client.

If ShadowsocksX-NG already provides HTTP proxy, why not use it?

**Comment by daciwei** [comment]
> This is incorrect. Lots of terminal utilities support SOCKS, e.g. curl

for example  docker needs HTTP proxy

> If ShadowsocksX-NG already provides HTTP proxy, why not use it?

I am using it now. 

But for some situation comand line client will be more usefull. Not all hardware has a GUI :)
Now for that kind situations in addition to SS I have to use polipo proxy to convert SOCKS to HTTP.

If you develop both server and **slient** side, HTTP proxy will be very handy, less additional apps required.

**Comment by riobard** [comment]
According to https://github.com/docker/docker/issues/5989#issuecomment-205393601 you can use SOCKS proxy with Docker. 

I guess the question is actually what applications only support HTTP proxy but not SOCKS proxy?

**Comment by daciwei** [comment]
`go get ...` not working with socks as I know... (or I don't know how) 

Docker on MacOs not using SOCK proxy (at least for me).
I saw this commit before and tried lots of times to make it work... but only HTTP(S)

Official docker app for mac even don't have other proxy options...

![screen shot 2017-04-11 at 1 13 12 pm](https://cloud.githubusercontent.com/assets/13871989/24906851/2e004ed2-1ec2-11e7-9446-e26e9aa61726.png)

**Comment by riobard** [comment]
What if you choose "Use system proxy" and set the system proxy to SOCKS? Will it work?

**Comment by daciwei** [comment]
unfortunately also not working.

<img src="https://cloud.githubusercontent.com/assets/415928/24956901/6fb4ab18-1f92-11e7-8f7f-ceb9a14b2b75.png"  width="70%">

<img src="https://cloud.githubusercontent.com/assets/415928/24957090/2ee63fe2-1f93-11e7-87fc-c458fdadfd58.png"  width="70%">

So I use HTTP proxy a lot... :)

**Comment by riobard** [comment]
I see. It seems HTTP proxy is still necessary. I'll consider adding it next.

**Comment by daciwei** [comment]
@riobard Yes, HTTP  will be very handy. I would  really appreciate if you can and it to development plans.

**Comment by bigtan** [comment]
https://github.com/bigtan/cow 

I just give go-ss2 a http proxy.

**Comment by nadoo** [comment]
@targence 
https://github.com/nadoo/glider

**Comment by helphi** [comment]
@riobard what's the progress？

**Comment by riobard** [comment]
Currently not planned.

**Comment by leewi9** [comment]
https://github.com/snail007/goproxy

maybe you can try this, or refer to this 

or this

**Comment by wanjochan** [comment]
anydesk need http proxy only

**Comment by JimLee1996** [comment]
> https://github.com/bigtan/cow
> 
> I just give go-ss2 a http proxy.

@bigtan Hi, I wanna implement a http listener in go-ss2 client too, and this repo is gone. Could you please re-open it for reference? Thank you.

**Comment by BigSully** [comment]
You can compile the http proxy support yourself with [this implementation](https://github.com/shadowsocks/go-shadowsocks2/issues/227).
